csi/crypto/crypto.go

```diff
@@ -97,8 +97,10 @@ func EncryptVolume(devicePath, passphrase string, cryptoParams *EncryptParams) e
 // OpenVolume opens volume so that it can be used by the client.
 func OpenVolume(volume, devicePath, passphrase string) error {
 	if isOpen, _ := IsDeviceOpen(VolumeMapper(volume)); isOpen {
-		logrus.Infof("Device %s is already opened at %s", devicePath, VolumeMapper(volume))
-		return nil
+		logrus.Infof("Device %s is already opened at %s, and close it first", devicePath, VolumeMapper(volume))
+		if err := CloseVolume(volume); err != nil {
+			return err
+		}
```

Comment:
Why do we need to close the volume here?

Reply:
In a normal process of attaching a volume via CSI, the volume on the worker node should be inactivated. (New deploying workload with PVC or scale the workload down and then scale the workload up)

In an abnormal process, the [NodeUnStageVolume](https://github.com/longhorn/longhorn-manager/blob/master/csi/node_server.go#L553) is unreachable and the encrypted volume is not closed, and then the new attaching request is coming.

Reply:
Or we can check if the encrypted volume is opened and then close it here.
https://github.com/longhorn/longhorn-manager/blob/master/csi/node_server.go#L126-L131

Reply:
I'm confused. The device is opened based on line 99. Isn't the volume open? Is it necessary to close it?

Reply:
Line 99 `IsDeviceOpen` only checks whether the device is opened.
The opening operation is the line 113 `nsexec.LuksOpen(...)`

Reply:
The original implementation continues staging the volume when the crypto device is already opened. This means the code will not reach `nsexec.LuksOpen()`, and the CSI then attempts to reuse the already opened crypto device. So I guess the question is, what is blocking the CSI from reusing the crypto device?

Reply:
Modified to check if the device path is null from the `cryptsetup status [device]` information before opening the device. If so, close the device before opening it.